### PR TITLE
support arbitrary metric logging from torchmetrics

### PR DESCRIPTION
**Pull request by sichu2023**
### Description
Training and validation `torchmetrics.Metric` can is organized by `TorchmetricsConfig`. This encapsulates metric class instantiation and naming through `get_metric_name` instead of using `field_factory`.

Currently model parallelism is not supported and will raise `NotImplementedError`.

### Type of changes
- [x]  New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/model.py [hunk 1]
```diff
@@ -531,6 +532,10 @@ class BioBertConfig(
     # loss reduction class
     loss_reduction_class: Type[MegatronLossType] = BERTMLMLossWithReduction
 
+    # metric logging
+    train_metric: Optional[MetricConfig] = None
+    valid_metric: Optional[MetricConfig] = None
```
Comment (farhadrgh) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/model.py:537]: Do you need to include these in `_OVERRIDE_BIONEMO_CONFIG_DEFAULTS` as well: 
https://github.com/NVIDIA/bionemo-framework/blob/9cec09f963cf5747193262b84bb700148280792b/sub-packages/bionemo-llm/src/bionemo/llm/model/config.py#L42

**Review by farhadrgh — APPROVED** [review]
I tentatively approve to unblock, but the tests are failing, and I wasn’t able to experiment with it.

File: sub-packages/bionemo-llm/src/bionemo/llm/model/config.py [hunk 2]
```diff
@@ -150,3 +152,31 @@ def override_mutate_possibly_extra_mutated_fiddle(
         target_cfg.set_hparam(f, source_cfg.get_hparam(f), also_change_value=False)
         # 2. Update the lazily untracked values (if the same variable name is used post-init)
         setattr(target_cfg, f, getattr(source_cfg, f))
+
+
+@dataclass
+class MetricConfig:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/config.py:158]: could you please add unit test for this class?

Comment (dorotat-nv) [reply sub-packages/bionemo-llm/src/bionemo/llm/model/config.py:158]: and maybe we can rename it to TorchMetricsConfig?

**Review by dorotat-nv — CHANGES_REQUESTED** [review]

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py [hunk 3]
```diff
@@ -317,11 +317,17 @@ def training_step(self, batch, batch_idx: Optional[int] = None) -> Tensor:
...
-        if self.train_ppl is not None:
+        if self.train_metric is not None:
             if self.is_on_logging_device():
-                self.train_ppl(logits, batch["labels"])
-
-            self.log("train_ppl", self.train_ppl, on_step=True, on_epoch=False, prog_bar=True)
+                self.train_metric(logits, batch["labels"])
+
+            self.log(
+                self.config.train_metric.get_metric_name("train"),
```
Comment (pstjohn) [inline sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:325]: `train_metric.get_metric_name("train")` is an odd API; what's going on here? wouldn't the metric name be something like "mse"? wouldn't we want to be logging all the train metrics?

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py [hunk 4]
```diff
@@ -259,9 +257,11 @@ def __init__(
         self._forward_step = forward_step
         self.model_transform = model_transform
 
-        # torchmetrics must init here for fiddle serialization
-        self.train_ppl = torchmetrics.text.Perplexity(ignore_index=MLM_LOSS_IGNORE_INDEX) if log_train_ppl else None
-        self.valid_ppl = torchmetrics.text.Perplexity(ignore_index=MLM_LOSS_IGNORE_INDEX) if log_val_ppl else None
+        # configure metrics
+        self.train_metric = self.config.train_metric.get_instance() if self.config.train_metric else None
+        self.valid_metric = self.config.valid_metric.get_instance() if self.config.valid_metric else None
+        if (self.train_metric or self.valid_metric) and not self.is_on_logging_device:
```
[the remark this answers is not in the document]
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:263]: Yes and this is an imperfect solution which might lock the other devices.

We can condition on tensor/pipeline model parallelism If `parallel_state` has the functionality to inspect the tensor/pipeline model parallel size even when tp=pp=1. `get_tensor_model_parallel_world_size` and `get_pipeline_model_parallel_world_size` will throw an error in such case.

[the remark this answers is not in the document]
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:263]: Would be great to have suggestions on alternatives.

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py (hunk 3 above)
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:325]: Basically a way to prepend "{stage}_" in front of your metric name. If no metric name is provided it will use the class name.

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py [hunk 5]
```diff
@@ -84,6 +84,7 @@ def test_esm2_finetune_token_classifier(
         assert weights_ckpt.is_dir()
         assert io.is_distributed_ckpt(weights_ckpt)
         assert simple_ft_metrics.collection_train["loss"][0] > simple_ft_metrics.collection_train["loss"][-1]
+        # assert trainer.logged_metrics["val_acc"].item() <= 0.5  # TODO @farhad for a reasonable value
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py:87]: @farhadrgh will have to pick your brain.

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py [hunk 6]
```diff
@@ -136,6 +137,7 @@ def test_esm2_finetune_regressor(
         assert weights_ckpt.is_dir()
         assert io.is_distributed_ckpt(weights_ckpt)
         assert simple_ft_metrics.collection_train["loss"][0] > simple_ft_metrics.collection_train["loss"][-1]
+        # assert trainer.logged_metrics["val_mse"].item() <= 0.5  # TODO @farhad for a reasonable value
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py:140]: @farhadrgh will have to pick your brain.

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py [hunk 7]
```diff
@@ -189,6 +191,7 @@ def test_esm2_finetune_classifier(
         assert weights_ckpt.is_dir()
         assert io.is_distributed_ckpt(weights_ckpt)
         assert simple_ft_metrics.collection_train["loss"][0] > simple_ft_metrics.collection_train["loss"][-1]
+        # assert trainer.logged_metrics["val_acc"].item() <= 0.5  # TODO @farhad for a reasonable value
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py:194]: @farhadrgh will have to pick your brain.

File: sub-packages/bionemo-llm/src/bionemo/llm/model/config.py [hunk 8]
```diff
@@ -150,3 +154,33 @@ def override_mutate_possibly_extra_mutated_fiddle(
         target_cfg.set_hparam(f, source_cfg.get_hparam(f), also_change_value=False)
         # 2. Update the lazily untracked values (if the same variable name is used post-init)
         setattr(target_cfg, f, getattr(source_cfg, f))
+
+
+@dataclass
+class TorchmetricsConfig:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/config.py:160]: why not to use pydantic?

Comment (dorotat-nv) [reply sub-packages/bionemo-llm/src/bionemo/llm/model/config.py:160]: I would reimplement it to pydantic and add some validations

ie

```
from pydantic import BaseModel, Field, model_validator

class TorchmetricsConfig(BaseModel):
    class_path: str
    task: Literal["lm", "classification", "regression"]
    metric_name: str
    kwargs: Optional[dict[str, Any]] = Field(default_factory=dict)
    
    @model_validator(mode='after')
    def validate_class_path(self) -> 'TorchmetricsConfig':
        try:
            self.get_instance()
        except (ImportError, AttributeError) as e:
            raise ValueError(f"Invalid class_path: {self.class_path}. Error: {str(e)}")
        return self
    
    def get_instance(self) -> torchmetrics.Metric:
        """Dynamically imports and instantiates the metric class."""
        if "." in self.class_path:
            module_path, class_name = self.class_path.rsplit(".", 1)
            module = importlib.import_module(f"torchmetrics.{module_path}")
        else:
            class_name = self.class_path
            module = importlib.import_module("torchmetrics")
            
        cls = getattr(module, class_name)
        return cls(**self.kwargs)
```

File: sub-packages/bionemo-llm/src/bionemo/llm/model/config.py [hunk 9]
```diff
@@ -150,3 +154,33 @@ def override_mutate_possibly_extra_mutated_fiddle(
...
+    """MetricConfig to instantiate torchmetrics.Metric class.
```
Comment (dorotat-nv) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/config.py:161]: please update docs

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dorotat-nv) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:258]: imho, this error should be thrown when the metric config is initialised.  In pydantic it could be

```
from pydantic import BaseModel, Field
 
class TorchmetricsConfig(BaseModel):
    class_path: str
    task: Literal["lm", "classification", "regression"]
    metric_name: str
    kwargs: Optional[dict[str, Any]] = Field(default_factory=dict)
    model_parallelism: bool = False
....
    def validate_parallel_logging(self) -> None:
        if self.model_parallelism:
            raise NotImplementedError(f"{self.__class__.__name__} logging is not implemented with model parallelism yet.")

```


and then you would need to intilised metrcis with 
```
valid_metric = TorchmetricsConfig(
            class_path="Accuracy",
            task="classification",
            kwargs={
                "task": "multiclass",
                "threshold": 0.5,
                "num_classes": data_module.train_dataset.label_tokenizer.vocab_size,
            },
            metric_name="val_acc",
            model_parallelism: any([p > 1 [tensor_model_parallel, pipeline_model_parallel]])
        )
```

alternatively you can add parallelism check to the config and pass pp and tp settings

**Review by dorotat-nv — CHANGES_REQUESTED** [review]

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:258]: Since model parallelism is supported at `BionemoLightningModule` level I prefer to move the model parallelism check outside of `TorchmetricsConfig`. I have moved that into the training and finetuning scripts.

File: sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py (hunk 5 above)
Comment (sichu2023) [reply sub-packages/bionemo-esm2/tests/bionemo/esm2/scripts/test_finetune_esm2.py:88]: Thanks! Feel free to push your commits directly.

File: sub-packages/bionemo-llm/src/bionemo/llm/model/config.py (hunk 8 above)
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/model/config.py:160]: Because we still have implementation beyond pydantic, e.g. `train_esm.py` and `finetune_esm.py` where most unittest and dev happen. That alone will be another discussion.

**Comment by sichu2023** [comment]
Discussed with @farhadrgh. Will create an [issue](https://github.com/NVIDIA/bionemo-framework/issues/684) for esm2 finetuning golden value.

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py [hunk 10]
```diff
@@ -312,26 +305,47 @@ def forward_step(self, batch) -> Tensor:
         assert self.module is not None
         return self._forward_step(self.module, batch)
 
+    def update_metric(self, batch, outputs, metric, task: Literal["lm", "classification", "regression"]) -> None:
```
Comment (dorotat-nv) [inline sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:308]: please add better docstring for this class. Also, what does "lm" means here?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 11]
```diff
@@ -276,6 +277,26 @@ def train_model(
...
+            task="classification",
+            kwargs={
+                "task": "multiclass",
+                "threshold": 0.5,
+                "num_classes": data_module.train_dataset.label_tokenizer.vocab_size,
+            },
+            metric_name="val_acc",
+        )
+
+    if tensor_model_parallel_size * pipeline_model_parallel_size > 1 and (
```
Comment (dorotat-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:295]: super!

**Review by dorotat-nv — APPROVED** [review]
Cosmetic changes related to naming and better docs, otherwise LGTM

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py (hunk 10 above)
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:308]: For pretraining the LM. Let me rename it into pretraining to clarify.

**Review by farhadrgh — APPROVED** [review]
